### DISCUSSION TITLE
Read from SQL during task

### DISCUSSION THREAD

**Discussion by steffejr** [jspsych/jsPsych#1951] [Q&A]
Hello all,
I would like to have a jsPsych task read from a MySQL database and I cannot figure out how to do it. Within my jsPsych code I would like to return data from MySQL that I can use within jsPsych. I have been able to read from the database and write the data via console.log() but not to get data as a variable.

Can anyone help please?  Thank you.

**Comment by becky-gilbert** [comment]
Hi @steffejr, if you're able to read from the database and write the data to the console, then you should be most of the way there! Can you share a bit of (simplified) code so that we can see where/how you're reading the data in, and give you some tips for how to turn that data into a variable that jsPsych can use? 

Without having seen exactly what you're doing, one approach that comes to mind would be to initialize a global variable to hold the data, and then use the [`call-function` plugin](https://www.jspsych.org/plugins/jspsych-call-function/) to access the MySQL data and save the result to this variable. You'll want to set the `async` parameter to `true` ([example here](https://www.jspsych.org/plugins/jspsych-call-function/#async-function-call)) in order to force jsPsych to wait for the data transfer to finish before moving on with the experiment. 

Another tip is to use [dynamic parameters](https://www.jspsych.org/overview/dynamic-parameters/) and [event-related callback functions](https://www.jspsych.org/overview/callbacks/) to access and manipulate the global variable that contains your database data, since it will be undefined when the page first loads.

Here's a sketch of the general approach:
```js
var mysql_data;

var get_data = {
    type: 'call-function',
    async: true,
    func: function(done) {
        // get data from database here and store in mysql_data variable
        mysql_data = result;
        done(); // call the callback function to end the call-function trial once the data transfer has finished
    }
}

var trial = {
    // other parameters ...
    stimulus: function() {
        return mysql_data.stimulus[0];
    }
}
```
Would something like this work?
